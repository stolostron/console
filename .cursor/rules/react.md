# React Rules for ACM Console

## Component Patterns

### Functional Components
- Use functional components with hooks exclusively
- Implement proper prop interfaces with TypeScript
- Use React.memo for expensive components
- Follow established naming conventions (PascalCase)

### Hook Usage
- Use custom hooks for business logic separation
- Implement proper dependency arrays in useEffect
- Use useCallback for event handlers passed as props
- Leverage useMemo for expensive computations

### State Management
- Use Recoil atoms through useSharedAtoms() pattern
- Access atoms with useRecoilValue() for reading
- Use useRecoilValueGetter() for async operations
- Follow established atom naming conventions

Example (based on actual codebase patterns):
```typescript
/* Copyright Contributors to the Open Cluster Management project */
import { useMemo } from 'react'
import { useRecoilValue } from 'recoil'
import { useTranslation } from '../lib/acm-i18next'
import { useSharedAtoms } from '../shared-recoil'
import { AcmTable } from '../ui-components'

export function ClusterList() {
  const { t } = useTranslation()
  const { managedClustersState } = useSharedAtoms()
  const clusters = useRecoilValue(managedClustersState)
  
  const filteredClusters = useMemo(
    () => clusters.filter(cluster => cluster.status?.conditions),
    [clusters]
  )
  
  return (
    <AcmTable
      items={filteredClusters}
      columns={clusterColumns}
      keyFn={cluster => cluster.metadata.uid}
    />
  )
}
```

### Conditional Rendering
- Use logical AND (`&&`) for simple conditional rendering (ACM standard pattern):
  ```typescript
  {condition && <Component />}
  {props.additionalLabels && (<span><AcmLabels /></span>)}
  {expandable && itemCount > columnCount && (<AcmButton />)}
  ```
- Use ternary operators for either/or scenarios:
  ```typescript
  {props.isCompact ? <Popover /> : renderLabelGroup()}
  {showAll ? t('Show less') : t('Show all')}
  ```
- Early returns for major conditional logic:
  ```typescript
  if (props.hidden) return <></>
  if (!data) return <LoadingState />
  ```

### Performance
- Use `React.memo` for components that receive the same props frequently
- Use `useMemo` for expensive calculations
- Use `useCallback` for functions passed to child components
- Implement lazy loading for heavy components
- Avoid inline object/array creation in render
- Avoid using index as the key prop value
- Avoid using inline functions in JSX or TSX blocks
- Use proper key props for list items

### Error Boundaries
- Implement error boundaries for route-level components
- Use proper error state management
- Provide user-friendly error messages
- Log errors for debugging purposes

### Accessibility
- Use semantic HTML elements
- Implement proper ARIA attributes
- Ensure keyboard navigation works
- Test with screen readers

### Component Organization
- Keep components small and focused on a single responsibility
- Use PascalCase for component names and component files
- Use descriptive component names that indicate the component's purpose
- Avoid abbreviations in component names
- Separate presentation from business logic
- Use proper file structure (components/, ui-components/, routes/)
- Export components properly with named exports
- Avoid props drilling where props are only passed down to child components
- Do not mutate props or state - create a copy array or object if possible
