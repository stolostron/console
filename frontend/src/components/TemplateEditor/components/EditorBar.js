/* Copyright Contributors to the Open Cluster Management project */
'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import { SearchInput } from '@patternfly/react-core'
import '../css/editor-bar.css'
import { CloseIcon, UndoIcon, RedoIcon, NextIcon, PreviousIcon } from '../icons/Icons'

class EditorButton extends React.Component {
  static propTypes = {
    button: PropTypes.object,
    command: PropTypes.string,
    handleClick: PropTypes.func,
  }

  handleClick = () => {
    const {
      command,
      button: { disabled },
    } = this.props
    if (!disabled) {
      document.activeElement.blur()
      this.props.handleClick(command)
    }
  }

  handleKeyPress = (e) => {
    if (e.key === 'Enter') {
      this.props.handleClick(this.props.command)
    }
  }

  renderIcon = (icon) => {
    switch (icon) {
      case 'close':
        return <CloseIcon />
      case 'undo':
        return <UndoIcon />
      case 'redo':
        return <RedoIcon />
      case 'next':
        return <NextIcon />
      case 'previous':
        return <PreviousIcon />
    }
  }

  render() {
    const {
      button: { disabled, tooltip, icon, spacer, command },
    } = this.props
    if (spacer) {
      return <div className="editor-bar-spacer" />
    } else {
      const classes = classNames({
        'editor-bar-button': true,
        [`${command}`]: true,
        disabled,
      })
      return (
        <div
          className={classes}
          tabIndex={0}
          role={'button'}
          aria-label={tooltip}
          title={tooltip}
          onClick={this.handleClick}
          onKeyPress={this.handleKeyPress}
        >
          {icon ? this.renderIcon(icon) : <div>{tooltip}</div>}
        </div>
      )
    }
  }
}

class EditorBar extends React.Component {
  static propTypes = {
    handleEditorCommand: PropTypes.func,
    handleSearchChange: PropTypes.func,
    hasRedo: PropTypes.bool,
    hasUndo: PropTypes.bool,
    i18n: PropTypes.func,
    title: PropTypes.string,
    type: PropTypes.string,
  }

  constructor(props) {
    super(props)
    this.state = {
      searchName: '',
    }
  }

  handleClick = (command) => {
    this.props.handleEditorCommand(command)
  }

  handleSearch = (searchName) => {
    this.props.handleSearchChange(searchName)
    this.setState({ searchName })
  }

  handleClear = () => {
    this.props.handleSearchChange('')
    this.setState({ searchName: '' })
  }

  render() {
    const { hasUndo, hasRedo, type, title, i18n } = this.props
    const { searchName } = this.state

    const undoButtons = [
      {
        command: 'undo',
        tooltip: i18n ? i18n('editor.bar.undo') : 'Undo',
        icon: 'undo',
        disabled: !hasUndo,
      },
      {
        command: 'redo',
        tooltip: i18n ? i18n('editor.bar.redo') : 'Redo',
        icon: 'redo',
        disabled: !hasRedo,
      },
    ]

    const nextButtons = [
      {
        command: 'previous',
        tooltip: i18n ? i18n('editor.bar.previous') : 'Previous',
        icon: 'previous',
        disabled: !searchName,
      },
      {
        command: 'next',
        tooltip: i18n ? i18n('editor.bar.next') : 'Next',
        icon: 'next',
        disabled: !searchName,
      },
    ]

    const resetButtons = [
      {
        command: 'restore',
        tooltip: i18n ? i18n('editor.bar.reset') : 'Reset',
        disabled: !hasUndo && !hasRedo,
      },
    ]

    const closeButtons = [
      {
        command: 'close',
        tooltip: i18n ? i18n('editor.bar.close') : 'Close',
        icon: 'close',
      },
    ]

    const searchTitle = i18n ? i18n('find.label') : 'Find'
    return (
      <div className="editor-bar">
        <div className="editor-bar-group">
          <div className="editor-bar-title">{title}</div>
        </div>
        <div className="editor-bar-group">
          <div className="editor-bar-toolbar">
            <div className="editor-bar-section">
              {resetButtons.map((button) => {
                const { command } = button
                return <EditorButton key={command} command={command} button={button} handleClick={this.handleClick} />
              })}
            </div>
            <div className="editor-bar-section">
              {undoButtons.map((button) => {
                const { command } = button
                return <EditorButton key={command} command={command} button={button} handleClick={this.handleClick} />
              })}
            </div>
            <div className="editor-bar-section">
              <div className="editor-bar-search" role="region" aria-label={searchTitle} id={searchTitle}>
                <SearchInput
                  id={`template-editor-search-${type}`}
                  value={searchName}
                  aria-label={searchTitle}
                  placeholder={searchTitle}
                  onChange={this.handleSearch}
                  onClear={this.handleClear}
                />
                {nextButtons.map((button) => {
                  const { command } = button
                  return <EditorButton key={command} command={command} button={button} handleClick={this.handleClick} />
                })}
              </div>
            </div>
          </div>
        </div>
        <div className="editor-bar-close">
          <div className="editor-bar-section">
            {closeButtons.map((button) => {
              const { command } = button
              return <EditorButton key={command} command={command} button={button} handleClick={this.handleClick} />
            })}
          </div>
        </div>
      </div>
    )
  }
}

export default EditorBar
