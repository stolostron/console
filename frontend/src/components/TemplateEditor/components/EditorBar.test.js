/* Copyright Contributors to the Open Cluster Management project */
'use strict'

import React from 'react'
import EditorBar from './EditorBar'
import { render } from '@testing-library/react'

describe('EditorBar component', () => {
    it('renders as expected', () => {
        const fn = jest.fn()
        const exceptions = [{ text: 'bad', row: 0 }]
        const Component = () => {
            return (
                <EditorBar
                    hasUndo={false}
                    hasRedo={true}
                    exceptions={exceptions}
                    handleEditorCommand={fn}
                    handleSearchChange={fn}
                    i18n={fn}
                />
            )
        }
        const { asFragment } = render(<Component />)
        expect(asFragment()).toMatchSnapshot()
    })
})
