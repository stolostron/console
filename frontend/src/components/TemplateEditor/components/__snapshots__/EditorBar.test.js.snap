// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorBar component renders as expected 1`] = `
<DocumentFragment>
  <div
    class="editor-bar"
  >
    <div
      class="editor-bar-group"
    >
      <div
        class="editor-bar-title"
      />
    </div>
    <div
      class="editor-bar-group"
    >
      <div
        class="editor-bar-toolbar"
      >
        <div
          class="editor-bar-section"
        >
          <div
            class="editor-bar-button restore"
            role="button"
            tabindex="0"
          >
            <div />
          </div>
        </div>
        <div
          class="editor-bar-section"
        >
          <div
            class="editor-bar-button undo disabled"
            role="button"
            tabindex="0"
          >
            <svg
              height="16px"
              role="presentation"
              version="1.1"
              viewBox="0 0 18 12"
              width="16px"
            >
              <g
                fill="none"
                fill-rule="evenodd"
                stroke="none"
                stroke-width="1"
              >
                <g
                  fill="#EDEDED"
                  transform="translate(-827.000000, -218.000000)"
                >
                  <g
                    transform="translate(827.000000, 218.000000)"
                  >
                    <path
                      d="M14,5.3 C15.5,5.3 16.8,6.6 16.8,8 C16.8,9.5 15.5,10.7 14,10.7 L10,10.7 L10,12 L14,12 C16.3,12 18,10.3 18,8 C18,5.8 16.2,4 14,4 L2.4,4 L4.9,1.5 L4,0.6 L0,4.6 L4,8.6 L4.9,7.7 L2.4,5.3 L14,5.3 Z"
                      id="Fill"
                    />
                  </g>
                </g>
              </g>
            </svg>
          </div>
          <div
            class="editor-bar-button redo"
            role="button"
            tabindex="0"
          >
            <svg
              height="16px"
              role="presentation"
              version="1.1"
              viewBox="0 0 18 12"
              width="16px"
            >
              <g
                fill="none"
                fill-rule="evenodd"
                stroke="none"
                stroke-width="1"
              >
                <g
                  fill="#EDEDED"
                  transform="translate(-857.000000, -218.000000)"
                >
                  <g
                    transform="translate(857.000000, 218.000000)"
                  >
                    <path
                      d="M4,5.3 C2.5,5.3 1.2,6.6 1.2,8 C1.2,9.5 2.5,10.7 4,10.7 L8,10.7 L8,12 L4,12 C1.7,12 0,10.3 0,8 C0,5.8 1.8,4 4,4 L15.6,4 L13.1,1.5 L14,0.6 L18,4.6 L14,8.6 L13.1,7.7 L15.6,5.2 L4,5.2 L4,5.3 Z"
                      id="Fill"
                    />
                  </g>
                </g>
              </g>
            </svg>
          </div>
        </div>
        <div
          class="editor-bar-section"
        >
          <div
            class="editor-bar-search"
            role="region"
          >
            <div
              class="pf-c-search-input"
              id="template-editor-search-undefined"
            >
              <div
                class="pf-c-input-group"
              >
                <div
                  class="pf-c-search-input__bar"
                >
                  <span
                    class="pf-c-search-input__text"
                  >
                    <span
                      class="pf-c-search-input__icon"
                    >
                      <svg
                        aria-hidden="true"
                        fill="currentColor"
                        height="1em"
                        role="img"
                        style="vertical-align: -0.125em;"
                        viewBox="0 0 512 512"
                        width="1em"
                      >
                        <path
                          d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                        />
                      </svg>
                    </span>
                    <input
                      aria-label="Search input"
                      class="pf-c-search-input__text-input"
                      value=""
                    />
                  </span>
                </div>
              </div>
            </div>
            <div
              class="editor-bar-button previous disabled"
              role="button"
              tabindex="0"
            >
              <svg
                height="16px"
                role="presentation"
                version="1.1"
                viewBox="0 0 18 18"
                width="16px"
              >
                <g
                  fill="none"
                  fill-rule="evenodd"
                  stroke="none"
                  stroke-width="1"
                >
                  <g
                    fill="#EDEDED"
                    transform="translate(-1225.000000, -215.000000)"
                  >
                    <path
                      d="M1234,216.5 C1238.14214,216.5 1241.5,219.857864 1241.5,224 C1241.5,228.142136 1238.14214,231.5 1234,231.5 C1229.85786,231.5 1226.5,228.142136 1226.5,224 C1226.5,219.857864 1229.85786,216.5 1234,216.5 L1234,216.5 Z M1234,215.25 C1229.16751,215.25 1225.25,219.167508 1225.25,224 C1225.25,228.832492 1229.16751,232.75 1234,232.75 C1238.83249,232.75 1242.75,228.832492 1242.75,224 C1242.75,219.167508 1238.83249,215.25 1234,215.25 Z M1230.35502,224.334055 L1235.91465,224.334055 L1233.00667,227.242032 L1233.89056,228.125915 L1238.30997,223.706498 L1233.89498,219.2915 L1233.00225,220.166545 L1235.91907,223.08336 L1230.35944,223.08336 L1230.35502,224.334055 L1230.35502,224.334055 Z"
                      id="Fill"
                      transform="translate(1234.000000, 224.000000) scale(-1, 1) translate(-1234.000000, -224.000000) "
                    />
                  </g>
                </g>
              </svg>
            </div>
            <div
              class="editor-bar-button next disabled"
              role="button"
              tabindex="0"
            >
              <svg
                height="16px"
                role="presentation"
                version="1.1"
                viewBox="0 0 18 18"
                width="16px"
              >
                <g
                  fill="none"
                  fill-rule="evenodd"
                  stroke="none"
                  stroke-width="1"
                >
                  <g
                    fill="#EDEDED"
                    transform="translate(-1255.000000, -215.000000)"
                  >
                    <path
                      d="M1264,216.5 C1268.14214,216.5 1271.5,219.857864 1271.5,224 C1271.5,228.142136 1268.14214,231.5 1264,231.5 C1259.85786,231.5 1256.5,228.142136 1256.5,224 C1256.5,219.857864 1259.85786,216.5 1264,216.5 L1264,216.5 Z M1264,215.25 C1259.16751,215.25 1255.25,219.167508 1255.25,224 C1255.25,228.832492 1259.16751,232.75 1264,232.75 C1268.83249,232.75 1272.75,228.832492 1272.75,224 C1272.75,219.167508 1268.83249,215.25 1264,215.25 Z M1260.35502,224.334055 L1265.91465,224.334055 L1263.00667,227.242032 L1263.89056,228.125915 L1268.30997,223.706498 L1263.89498,219.2915 L1263.00225,220.166545 L1265.91907,223.08336 L1260.35944,223.08336 L1260.35502,224.334055 L1260.35502,224.334055 Z"
                      id="Fill"
                    />
                  </g>
                </g>
              </svg>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="editor-bar-close"
    >
      <div
        class="editor-bar-section"
      >
        <div
          class="editor-bar-button close"
          role="button"
          tabindex="0"
        >
          <svg
            height="16px"
            role="presentation"
            version="1.1"
            viewBox="-4 -4 16 16"
            width="16px"
          >
            <g>
              <path
                d="M9.8 8.6L8.4 10 5 6.4 1.4 10 0 8.6 3.6 5 .1 1.4 1.5 0 5 3.6 8.6 0 10 1.4 6.4 5z"
              />
            </g>
          </svg>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
